### PR TITLE
[mbedtls] Adding support to send larger payload

### PR DESCRIPTION
**Pull request by hihigupt**
Currently mbedtlsio does not allow to send larger payload then
framgent size. Adding support to allow sending larger application
payload.

### CODE REVIEW

**Comment by alexmrtn** [comment]
Tested this fix and it does solve the 16kB limit! Thanks

File: adapters/tlsio_mbedtls.c [hunk 1]
```diff
@@ -511,7 +543,7 @@ CONCRETE_IO_HANDLE tlsio_mbedtls_create(void *io_create_parameters)
                 {
                     result->tls_status = TLS_STATE_NOT_INITIALIZED;
                     mbedtls_init((void*)result);
-
+                    memset((void*)&result->send_complete_info, 0, sizeof(result->send_complete_info));
```
Comment (danewalton) [inline adapters/tlsio_mbedtls.c]: this seems to be doing duplicate work that `calloc` is doing

File: adapters/tlsio_mbedtls.c [hunk 2]
```diff
@@ -367,7 +387,19 @@ static int on_io_send(void *context, const unsigned char *buf, size_t sz)
     else
     {
         TLS_IO_INSTANCE *tls_io_instance = (TLS_IO_INSTANCE *)context;
-        if (xio_send(tls_io_instance->socket_io, buf, sz, on_send_complete, tls_io_instance) != 0)
+        ON_SEND_COMPLETE on_complete_callback = NULL;
+        void *context = NULL;
+
+        // Only allow Application data type message to send on_send_complete callback.
+        // In case when sending more then max_fregment size trigger only on when last data
```
Comment (danewalton) [inline adapters/tlsio_mbedtls.c]: nit: "then" should be "than" and "max_fregment" -> "max_fragment"

File: adapters/tlsio_mbedtls.c [hunk 3]
```diff
@@ -43,7 +42,7 @@ typedef enum TLSIO_STATE_ENUM_TAG
 
 typedef struct SEND_COMPLETE_INFO_TAG
 {
-    int send_complete_count;
+    int is_fragmented_req;
```
Comment (ewertons) [inline adapters/tlsio_mbedtls.c]: Nit: you can use bool if needed. Stdbool is already included.

File: adapters/tlsio_mbedtls.c (hunk 1 above)
Comment (hihigupt) [reply adapters/tlsio_mbedtls.c]: will remove.

---
In reply to: [354466164](https://github.com/Azure/azure-c-shared-utility/pull/421#discussion_r354466164) [](ancestors = 354466164)

File: adapters/tlsio_mbedtls.c (hunk 2 above)
Comment (hihigupt) [reply adapters/tlsio_mbedtls.c]: Ack

---
In reply to: [354466478](https://github.com/Azure/azure-c-shared-utility/pull/421#discussion_r354466478) [](ancestors = 354466478)

File: adapters/tlsio_mbedtls.c [hunk 4]
```diff
@@ -687,19 +719,30 @@ int tlsio_mbedtls_send(CONCRETE_IO_HANDLE tls_io, const void *buffer, size_t siz
...
-            }
-            else
+            int out_left = size;
+            result = 0;
+
+            do
             {
-                result = 0;
-            }
+                tls_io_instance->send_complete_info.is_fragmented_req = is_fragmented_send_request(tls_io_instance, out_left);
```
Comment (danewalton) [inline adapters/tlsio_mbedtls.c]: does this need to be called every iteration of the loop? pardon my ignorance here but this seems to be able to be set once?

Comment (hihigupt) [reply adapters/tlsio_mbedtls.c]: yes

File: adapters/tlsio_mbedtls.c (hunk 3 above)
Comment (hihigupt) [reply adapters/tlsio_mbedtls.c]: updated to bool

---
In reply to: [354467808](https://github.com/Azure/azure-c-shared-utility/pull/421#discussion_r354467808) [](ancestors = 354467808)

File: tests/tlsio_mbedtls_ut/tlsio_mbedtls_ut.c [hunk 5]
```diff
@@ -803,6 +861,93 @@ BEGIN_TEST_SUITE(tlsio_mbedtls_ut)
...
+        //assert
+        ASSERT_ARE_EQUAL(int, 0, result);
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+        //cleanup
+        (void)tlsio_mbedtls_close(handle, on_io_close_complete, NULL);
+        tlsio_mbedtls_destroy(handle);
+    }
+
+    TEST_FUNCTION(tlsio_mbedtls_send_large_payload_faliure)
```
Comment (ewertons) [inline tests/tlsio_mbedtls_ut/tlsio_mbedtls_ut.c]: typo on "faliure"

Comment (ewertons) [reply tests/tlsio_mbedtls_ut/tlsio_mbedtls_ut.c]: Thanks for the test!

**Pull request merged**